hack/generator/pkg/codegen/pipeline/unroll_recursive_types.go

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package pipeline
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	"k8s.io/klog/v2"
+
+	"github.com/Azure/azure-service-operator/hack/generator/pkg/astmodel"
+)
+
+const UnrollRecursiveTypesStageID = "unrollRecursiveTypes"
+
+// UnrollRecursiveTypes finds types that reference themselves and "unrolls" the reference. So a type that looks like:
+// type Error struct {
+//     code    string
+//     message string
+//     errors []Error
+// }
+// gets unrolled to look like:
+// type Error struct {
+//     code    string
+//     message string
+//     errors []Error_Unrolled
+// }
+// where Error_Unrolled looks like:
+// type Error_Unrolled struct {
+//     code    string
+//     message string
+// }
+// This usually happens for Error types (CloudError and the like). The recursive references must be removed because
+// controller-tools doesn't support generating "references" (JSON $ref) so it can't support recursive types today.
+// See https://github.com/kubernetes-sigs/controller-tools/issues/489 for more information.
+// Unrolling these types, while required for controller-tools to function, means
+// THAT THE TYPES WE GENERATE DON'T EXACTLY CONFORM TO THE PUBLISHED API!
+// In practice, we think this is ok because the types that reference themselves only return a depth of 1.
+// If we were to unroll all loops (rather than just types that directly reference themselves like we're doing here) that
+// "in practice" observation may no longer hold, so we avoid doing it here (it's also more complicated to do that).
+// TODO: We could require you to name the type in config or code in order to get it unrolled? There are currently 43 types being unrolled by this
```

Comment:
I'm definitely not opposed to changing the config we have. In some ways it feels like there are parts that would be better split apart? The Swagger repo has a config per-service type (so one for `microsoft.network`, one for `microsoft.compute`, etc). That keeps the configs somewhat small even though there are tons and tons of services.

In any case, a discussion for someplace other than here =)

Did you have any thoughts on the question I asked above?

Reply:
I'm happy with the approach you've taken here.

Reply:
Discussed this and decided to only do this for things with `Error` in the name, and also log an error or warning if we run into one that doesn't match that description.

Reply:
From discussion in chat it sounds like the plan is to do it automatically for any directly self-referential  type with "Error" in the name, and then raise an error for any other directly self-referential type? Or actually, if they don't get fixed automatically will we get a crash later on? I guess it would still be useful to fail fast at this point rather than getting a much more obscure error later.

That sounds good to me.

Reply:
This worked out quite well. If I run into a type without "error" in its name, I specifically raise an error. Turns out there are none currently (all 43 types that do this have Error in their name), which is exactly what we were going for.